rts/motoko-rts/src/gc/mark_compact.rs

```diff
@@ -123,87 +126,89 @@ unsafe fn mark_object<M: Memory>(mem: &mut M, obj: SkewedPtr, heap_base: u32) {
     }
 
     set_bit(obj_idx);
-    push_mark_stack(mem, obj as usize);
+    push_mark_stack(mem, obj as usize, obj_tag);
 }
 
 unsafe fn mark_stack<M: Memory>(mem: &mut M, heap_base: u32) {
-    while let Some(obj) = pop_mark_stack() {
-        mark_fields(mem, obj as *mut Obj, heap_base);
+    while let Some((obj, tag)) = pop_mark_stack() {
+        mark_fields(mem, obj as *mut Obj, tag, heap_base);
     }
 }
 
-unsafe fn mark_fields<M: Memory>(mem: &mut M, obj: *mut Obj, heap_base: u32) {
-    visit_pointer_fields(obj, heap_base as usize, |field_addr| {
-        mark_object(mem, *field_addr, heap_base);
-    });
-}
+unsafe fn mark_fields<M: Memory>(mem: &mut M, obj: *mut Obj, obj_tag: Tag, heap_base: u32) {
+    visit_pointer_fields(obj, obj_tag, heap_base as usize, |field_addr| {
+        let field_value = *field_addr;
+        mark_object(mem, field_value, heap_base);
 
-unsafe fn thread_roots(static_roots: SkewedPtr, heap_base: u32) {
-    // Static roots
-    let root_array = static_roots.as_array();
-    for i in 0..root_array.len() {
-        thread_obj_fields(root_array.get(i).unskew() as *mut Obj, heap_base);
-    }
-    // No need to thread closure table here as it's on heap and we already marked it
+        // Thread if backwards pointer
+        if field_value.unskew() < field_addr as usize {
```

Comment:
```suggestion
        if field_value.unskew() < field_addr as usize {
```
I guess that works, but shouldn't this really be tesing againt the address of the header, not some field offset from the hearder? At least morally?

Reply:
visit_pointer_fields only visits unmarked fields, right? So there is no danger of threading the same field twice?

Reply:
You're right. I updated this code to compare the field value (pointer) with the object address, instead of field address.